rgdps/api/commands/framework.py

```diff
@@ -0,0 +1,533 @@
+    def is_level(self) -> bool:
+        return self.level is not None
+
+    @property
+    def is_message(self) -> bool:
+        return self.target_user is not None
+
+    # Implementation of context requirements
+    @property
+    def mysql(self) -> AbstractMySQLService:
+        return self._base_context.mysql
+
+    @property
+    def redis(self) -> Redis:
+        return self._base_context.redis
+
+    @property
+    def meili(self) -> MeiliClient:
+        return self._base_context.meili
+
+    @property
+    def storage(self) -> AbstractStorage:
+        return self._base_context.storage
+
+    @property
+    def user_cache(self) -> AbstractAsyncCache[User]:
+        return self._base_context.user_cache
+
+    @property
+    def password_cache(self) -> AbstractAsyncCache[str]:
+        return self._base_context.password_cache
+
+    @property
+    def http(self) -> httpx.AsyncClient:
+        return self._base_context.http
+
+
+class CommandRoutable(ABC):
+    """An abstract class defining interfaces that `CommandRouter` can use to
+    execute commands."""
+
+    name: str
+
+    @abstractmethod
+    async def execute(self, ctx: CommandContext) -> str:
+        ...
+
+
+class CommandRouter:
+    """An (optionally) inheritable command router class, responible for
+    registering and executing commands handlers. Supports nesting routers
+    for command groups."""
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+        self._routes: dict[str, CommandRoutable] = {}
+
+    def register_command(self, command: CommandRoutable) -> None:
+        self._routes[command.name] = command
+
+    def merge(self, router: CommandRouter) -> None:
+        """Combines the routes of two routers, overriding old entries
+        with new ones (raises a warning)."""
+
+        for key, value in router._routes.items():
+            if key in self._routes:
+                logger.warning(
+                    "Command router merge has overwritten an existing command!",
+                    extra={
+                        "command": key,
+                        "source": router.name,
+                        "target": self.name,
+                    },
+                )
+            self._routes[key] = value
+
+    def register(self) -> Callable[[CommandRoutable], CommandRoutable]:
+        """A decorator version of `register_command`."""
+
+        def decorator(command: CommandRoutable) -> CommandRoutable:
+            self.register_command(command)
+            return command
+
+        return decorator
+
+    async def entrypoint(
+        self,
+        command: str,
+        user: User,
+        base_ctx: Context,
+        level: Level | None = None,
+        target_user: User | None = None,
+    ) -> str:
+        """Defines an entrypoint for command execution. This method is
+        supposed to be the first one called when executing a command."""
+
+        try:
+            parsed_params = _parse_params(command)
```

Comment:
maybe `_parse_and_resolve_params()`? it reads very low level/unimportant rn but it's doing a lot

Reply:
This isn't actually responsible for resolving params (thats the job of `_parse_to_type` which yeah I do agree could use a rename which I'll address on the next commit)

Reply:
Oh sorry, I didn't realize there was a function in the module named `_parse_params`, I was referring to the methods,

https://github.com/RealistikDash/RealistikGDPS/pull/37/files#diff-38ca8201f7b2870813d2cae6222e30c03bb441cdc8714720ae1992322016bc55R509-R512